### ISSUE TITLE
Incorrect diagnostic: error CS0648: '' is a type not supported by the language (tuple decoding issue)

### ISSUE THREAD

**Issue by ltrzesniewski** [dotnet/roslyn#21727]
**Version Used**:

- Visual Studio 15.3.2
- Roslyn 2.3.2.62019 (b7354608)

**Steps to Reproduce**:

- Create a new solution with two *Class Library (.NET Framerwork)* projects ProjectA and ProjectB
- Make sure the target framework of both projects is set to a .NET version prior to 4.7, the default v4.5.2 will do
- Add the `System.ValueTuple` NuGet package to ProjectA
- Reference ProjectA from ProjectB
- Add the following class to ProjectA:

```C#
using System.Collections;
using System.Collections.Generic;

namespace ProjectA
{
    public class ClassA : IEnumerable<(int foo, int bar)>
    {
        private readonly List<(int foo, int bar)> _values = new List<(int, int)>();

        public void Add(int foo, int bar) => _values.Add((foo, bar));

        public IEnumerator<(int foo, int bar)> GetEnumerator() => _values.GetEnumerator();
        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
    }
}
```

- Add the following class to ProjectB:

```C#
using ProjectA;

namespace ProjectB
{
    public class ClassB
    {
        public ClassB()
        {
            new ClassA { { 42, 10 } };
        }
    }
}
```

- Compile

**Expected Behavior**:

The compilation of ProjectB should fail with an error message indicating that a reference to `System.ValueTuple` is required:

> error CS0012: The type 'ValueTuple<,>' is defined in an assembly that is not referenced. You must add a reference to assembly 'System.ValueTuple, Version=4.0.2.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'.

Alternatively, the compilation could succeed, as we're *not* using the `GetEnumerator` function in ProjectB.

**Actual Behavior**:

The compilation fails on the `new ClassA { { 42, 10 } };` statement with an unhelpful/incomplete error message:

> error CS0648: '' is a type not supported by the language

Not only [this error](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0648) looks like a bad fit for this situation, its *type* argument is missing, which makes it quite puzzling.

**Comment by jcouv** [comment]
Thanks for reporting this.

I'll need to investigate more. There is a problem decoding tuple element names (` DecodeTupleTypesInternal`) in the second compilation. The encoding and decoding don't agree and I'm not yet sure which one is wrong.

Here's the repro I have so far:
```C#
        [Fact]
        [WorkItem(21727, "https://github.com/dotnet/roslyn/issues/21727")]
        public void Test()
        {
            var lib_cs = @"
using System.Collections;
using System.Collections.Generic;

public class ClassA : IEnumerable<(int alice, int bob)>
{
    private readonly List<(int alice, int bob)> _values = new List<(int, int)>();

    public void Add(int alice, int bob) => _values.Add((alice, bob));

    public IEnumerator<(int alice, int bob)> GetEnumerator() => _values.GetEnumerator();
    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
}";
            var lib = CreateCompilation(lib_cs, references: new[] { MscorlibRef, ValueTupleRef, SystemRuntimeFacadeRef });
            lib.VerifyDiagnostics();

            var client_cs = @"
public class ClassB
{
    public ClassB()
    {
        new ClassA { { 42, 10 } };
    }
}";
            var comp = CreateCompilation(client_cs, references: new[] { MscorlibRef, lib.EmitToImageReference() });
            comp.VerifyDiagnostics(
                // (6,20): error CS0648: '' is a type not supported by the language
                //         new ClassA { { 42, 10 } };
                Diagnostic(ErrorCode.ERR_BogusType, "{ { 42, 10 } }").WithArguments("").WithLocation(6, 20)
                );
        }
```

**Comment by jcouv** [comment]
Fix is merged to master branch, which I expect to ship with 15.8.
